lib/dataset/extract.cpp

```diff
@@ -0,0 +1,143 @@
+  } else if (data.dtype() == dtype<bucket<DataArray>>) {
+    const auto &[i, dim, buf] = data.constituents<DataArray>();
+    return copy_ranges_from_buffer(indices, dim, buf);
+  } else {
+    const auto &[i, dim, buf] = data.constituents<Dataset>();
+    return copy_ranges_from_buffer(indices, dim, buf);
+  }
+}
+
+Variable dense_or_bin_indices(const Variable &var) {
+  return is_bins(var) ? var.bin_indices() : var;
+}
+
+Variable dense_or_copy_bin_elements(const Variable &dense_or_indices,
+                                    const Variable &data) {
+  return is_bins(data) ? copy_ranges_from_bins_buffer(dense_or_indices, data)
+                       : dense_or_indices;
+}
+} // namespace
+
+template <class T>
+T extract_ranges(const Variable &indices, const T &data, const Dim dim) {
+  T no_edges;
+  if constexpr (std::is_same_v<T, Variable>)
+    no_edges = data;
+  else
+    no_edges = strip_edges_along(data, dim);
+  // 1. Operate on dense data, or equivalent array of indices (if binned) to
+  // obtain output data of correct shape with proper meta data.
+  auto dense = transform_data(no_edges, dense_or_bin_indices);
+  auto out =
+      copy_ranges_from_buffer(indices, dim, dense).template bin_buffer<T>();
+  // 2. If we have binned data then the data of the DataArray or Dataset
+  // obtained in step 1. give the indices into the underlying buffer to be
+  // copied. This then replaces the data to obtain the final result. Does
+  // nothing if dense data.
+  return transform_data(out, dense_or_copy_bin_elements, no_edges);
+}
+
+namespace {
+template <class T> T extract_impl(const T &obj, const Variable &condition) {
+  if (condition.dtype() != dtype<bool>)
+    throw except::TypeError(
+        "Cannot extract elements based on condition with non-boolean dtype. If "
+        "you intended to select a range based on a label you must specify the "
+        "dimension.");
+  if (condition.dims().ndim() != 1)
+    throw except::DimensionError("Condition must by 1-D, but got " +
+                                 to_string(condition.dims()) + '.');
+  if (!obj.dims().includes(condition.dims()))
+    throw except::DimensionError(
+        "Condition dimensions " + to_string(condition.dims()) +
+        " must be be included in the dimensions of the sliced object " +
+        to_string(obj.dims()) + '.');
+
+  auto values = condition.values<bool>().as_span();
+  std::vector<scipp::index_pair> indices;
+  for (scipp::index i = 0; i < scipp::size(values); ++i) {
+    if (i > 0 && values[i - 1] == values[i])
+      continue;    // not an edge
+    if (values[i]) // rising edge
+      indices.emplace_back(i, scipp::size(values));
+    else if (i != 0) // falling edge
+      indices.back().second = i;
+  }
+  return extract_ranges(makeVariable<scipp::index_pair>(Dims{condition.dim()},
+                                                        Shape{indices.size()},
+                                                        Values(indices)),
```

Comment:
Can it take advantage of this?
```suggestion
                                                        Values(std::move(indices))),
```

Reply:
No it cannot, since `Variable` is not based on `std::vector`.

Reply:
We could use `element_array`, but then we would need to do two passes to first determine the size, i.e., more code, i.e., premature optimization.